rllib/models/specs/specs_base.py

```diff
@@ -0,0 +1,209 @@
+    ) -> None:
+        self._expected_shape = self._parse_expected_shape(shape, shape_vals)
+        self._dtype = dtype
+
+    @property
+    def shape(self) -> Tuple[int]:
+        """Returns a `tuple` specifying the tensor shape."""
+        return self._expected_shape
+
+    @property
+    def dtype(self) -> Any:
+        """Returns a dtype specifying the tensor dtype."""
+        return self._dtype
+
+    @DeveloperAPI
+    def validate(self, tensor: TensorType) -> None:
+        """Checks if the shape and dtype of the tensor matches the specification.
+
+        Args:
+            tensor: The tensor to be validated.
+
+        Raises:
+            ValueError: If the shape or dtype of the tensor does not match the
+        """
+        shape = self.get_shape(tensor)
+        if len(shape) != len(self._expected_shape):
+            raise ValueError(_INVALID_SHAPE.format(self._expected_shape, shape))
+
+        for expected_d, actual_d in zip(self._expected_shape, shape):
+            if isinstance(expected_d, int) and expected_d != actual_d:
+                raise ValueError(_INVALID_SHAPE.format(self._expected_shape, shape))
+
+        dtype = self.get_dtype(tensor)
+        if self.dtype and dtype != self.dtype:
+            raise ValueError(_INVALID_DTYPE.format(self.dtype, tensor.dtype))
+
+    @abc.abstractmethod
+    def get_shape(self, tensor: TensorType) -> Tuple[int]:
+        """Returns the shape of a tensor.
+
+        Args:
+            tensor: The tensor whose shape is to be returned.
+
+        Returns:
+            A `tuple` specifying the shape of the tensor.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def get_dtype(self, tensor: TensorType) -> Any:
+        """Returns the data type of a tensor.
+
+        Args:
+            tensor: The tensor whose data type is to be returned.
+
+        Returns:
+            The data type of the tensor.
+        """
+        raise NotImplementedError
+
+    @DeveloperAPI
+    def sample(self, fill_value: Union[float, int] = 0) -> TensorType:
+        """Samples a tensor with the specified value that is matches the specs.
+
+        Args:
+            fill_value: The value to fill the tensor with.
+
+        Returns:
+            A tensor with the specified value that matches the specs.
+        """
+        sampled_shape = self._sample_shape()
+        return self._sample(sampled_shape, fill_value)
+
+    @abc.abstractmethod
+    def _sample(
+        self, shape: Tuple[int], fill_value: Union[float, int] = 0
+    ) -> TensorType:
+        """Returns a sample tensor with the given shape.
+
+        Args:
+            shape: The shape of the tensor to be sampled.
+            fill_value: The value to fill the tensor with.
+
+        Returns:
+            A tensor with the specified value that matches the specs.
+        """
+        raise NotImplementedError
+
+    def _sample_shape(self) -> Tuple[int]:
```

Comment:
tensor shape not sample shape

Reply:
This is sample_shape because of the reason I mentioned above.